lib/api/multiObjectDelete.js

```diff
@@ -118,58 +195,74 @@ function _parseXml(xmlToParse, next) {
-                            key,
-                            error: err,
-                        });
+    return async.forEachLimit(inPlay, 5, (entry, moveOn) => {
+        async.waterfall([
+            // for obj deletes, no need to check acl's at object level
+            // (authority is at the bucket level for obj deletes)
+            callback => metadataGetObject(bucketName, entry.key,
+                entry.versionId, log, (err, objMD) => {
+                    // if general error from metadata return error
+                    if (err && !err.NoSuchKey) {
+                        log.error('error getting object MD',
+                                { error: err, key: entry.key });
+                        errorResults.push({ entry, error: err });
                         return moveOn();
                     }
-                    if (objMD['content-length']) {
-                        totalContentLengthDeleted +=
-                            objMD['content-length'];
+                    // if particular key does not exist, AWS returns success
+                    // for key so add to successfullyDeleted list and move on
+                    if (err && err.NoSuchKey) {
+                        successfullyDeleted.push({ entry });
+                        return moveOn();
                     }
-                    numOfObjects++;
-                    successfullyDeleted.push(key);
-                    return moveOn();
-                });
+                    return callback(null, objMD);
+                }),
+            (objMD, callback) => preprocessingVersioningDelete(bucketName,
+                bucket, entry.key, objMD, entry.versionId, log,
+                (err, options) => callback(err, options, objMD)),
+            (options, objMD, callback) => {
+                const deleteInfo = {};
+                if (options && options.deleteData) {
+                    deleteInfo.deleted = true;
+                    return services.deleteObject(bucketName, objMD,
+                        entry.key, options, log, err =>
+                        callback(err, objMD, deleteInfo));
+                }
+                deleteInfo.newDeleteMarker = true;
+                // This call will create a delete-marker
+                return createAndStoreObject(bucketName, bucket, entry.key,
+                    objMD, authInfo, canonicalID, null, request,
+                    deleteInfo.newDeleteMarker, null, log, err =>
+                    callback(err, objMD, deleteInfo));
+            },
+        ], (err, objMD, deleteInfo) => {
+            if (err) {
+                log.error('error deleting object', { error: err, entry });
+                errorResults.push({ entry, error: err });
+                return moveOn();
+            }
+            if (deleteInfo.deleted && objMD['content-length']) {
+                totalContentLengthDeleted += objMD['content-length'];
+            }
+            numOfObjects++;
```

Comment:
`numOfObjects` which represents number of objects deleted, should be incremented only when a delete marker is created for the first time for the object.